pandas/core/indexes/base.py

```diff
@@ -4713,13 +4713,13 @@ def set_value(self, arr, key, value):
     @Appender(_index_shared_docs["get_indexer_non_unique"] % _index_doc_kwargs)
     def get_indexer_non_unique(self, target):
         target = ensure_index(target)
-        if is_categorical(target):
-            target = target.astype(target.dtype.categories.dtype)
         pself, ptarget = self._maybe_promote(target)
         if pself is not self or ptarget is not target:
             return pself.get_indexer_non_unique(ptarget)
 
-        if self.is_all_dates:
+        if is_categorical(target):
+            tgt_values = target.__array__()
```

Comment:
@TomAugspurger thoughts on this approach?

Reply:
You shouldn't call `__array__` directly. You should use `np.asarray` instead if you need an ndarray.